src/test/app/SetRegularKey_test.cpp

```diff
@@ -27,76 +27,106 @@ class SetRegularKey_test : public beast::unit_test::suite
+
+            env(fclear(alice, asfDisableMaster), sig(bob));
+            env(noop(alice), sig(bob));
+            env(noop(alice), sig(alice));
+
+            testcase("Revoke regular key");
+            env(regkey(alice, disabled));
+            env(noop(alice), sig(bob), ter(tefBAD_AUTH_MASTER));
+            env(noop(alice), sig(alice));
+        }
+
+        {
+            // See https://ripplelabs.atlassian.net/browse/RIPD-1721.
+            testcase("Set regular key to master key (before fix1721)");
+            Env env {*this, supported_amendments() - fix1721};
+            Account const alice("alice");
+            env.fund(XRP(10000), alice);
+
+            // Must be possible unless amendment `fix1721` enabled.
+            log << "set regular key to master key";
+            env(regkey(alice, alice), sig(alice));
+            log << "disable master key";
+            env(fset(alice, asfDisableMaster), sig(alice));
+            // No way to sign.
+            log << "implicitly sign transaction with master key";
+            env(noop(alice), ter(tefMASTER_DISABLED));
+            env(noop(alice), sig(alice), ter(tefMASTER_DISABLED));
+        }
+
+        {
+            testcase("Set regular key to master key (after fix1721)");
+            Env env {*this, supported_amendments() | fix1721};
+            Account const alice("alice");
+            env.fund(XRP(10000), alice);
+
+            // Must be possible unless amendment `fix1721` enabled.
+            env(regkey(alice, alice), ter(temCANT_USE_MASTER_KEY));
+        }
+
+        {
+            testcase("Password spent");
+            Env env(*this);
+            Account const alice("alice");
+            Account const bob("bob");
+            env.fund(XRP(10000), alice, bob);
+
+            auto ar = env.le(alice);
+            BEAST_EXPECT(ar->isFieldPresent(sfFlags) && ((ar->getFieldU32(sfFlags) & lsfPasswordSpent) == 0));
+
+            env(regkey(alice, bob), sig(alice), fee(0));
+
+            ar = env.le(alice);
+            BEAST_EXPECT(ar->isFieldPresent(sfFlags) && ((ar->getFieldU32(sfFlags) & lsfPasswordSpent) == lsfPasswordSpent));
+
+            // The second SetRegularKey transaction with Fee=0 should fail.
+            env(regkey(alice, bob), sig(alice), fee(0), ter(telINSUF_FEE_P));
+
+            env.trust(bob["USD"](1), alice);
+            env(pay(bob, alice, bob["USD"](1)));
+            ar = env.le(alice);
+            BEAST_EXPECT(ar->isFieldPresent(sfFlags) && ((ar->getFieldU32(sfFlags) & lsfPasswordSpent) == 0));
+        }
+
+        {
+            testcase("Universal mask");
+            Env env(*this);
+            Account const alice("alice");
+            Account const bob("bob");
+            env.fund(XRP(10000), alice, bob);
+
+            auto jv = regkey(alice, bob);
+            jv[sfFlags.fieldName] = tfUniversalMask;
+            env(jv, ter(temINVALID_FLAG));
+        }
```

Comment:
Why did you roll the separate test functions into `run`? Having the functions divided up helps with readability and maintainability.

Reply:
I disagree. I came into this class thinking adding a method would add a test. No, you must add a call to that method. Adding a method does not give the test a name either, though you have to give the method a name. Methods are not run independently, with separate setup and teardown. This class tries to look like other test frameworks, but it falls to live up to the expectations set by those frameworks. Instead, I'd rather it look exactly like what it's doing, so that adding to it is straightforward and intuitive.

Reply:
I have side with Ed on this one.  I hear that you have expectations based on other test frameworks that you've used.  And, believe me, I sympathize that the framework we're using could be better.  But there are a ton of tests already in this code base that follow the model that you just unrolled.  I believe, at least, the tests within this code base should be self consistent.

If you want to argue that _all_ of the tests in the code base should be unrolled you're welcome to make that argument to the team and provide a plan for making all the tests consistent with your vision.  If the team agrees with your vision then charge ahead.  But flattening just one test only adds to confusion.

Just my opinion.

Reply:
The tests are already inconsistent. Here are some that have a fully unrolled `run()` matching the pattern I use here: [`Buffer_test`](https://github.com/ripple/rippled/blob/develop/src/test/basics/Buffer_test.cpp#L38), [`TaggedCache_test`](https://github.com/ripple/rippled/blob/develop/src/test/basics/TaggedCache_test.cpp#L41), [`Slice_test`](https://github.com/ripple/rippled/blob/develop/src/test/basics/Slice_test.cpp#L30), [`Scheduler_test`](https://github.com/ripple/rippled/blob/develop/src/test/csf/Scheduler_test.cpp#L32), [`LedgerReplay_test`](https://github.com/ripple/rippled/blob/develop/src/test/app/LedgerReplay_test.cpp#L30). One surprising case is [`Manifest_test`](https://github.com/ripple/rippled/blob/develop/src/test/app/Manifest_test.cpp#L653), which has a mix of the two patterns. Given that both patterns are in common use, I'd like to tip the scales in favor of the unroll. I'll hold a quick poll in chat to see which way the winds blow, and if I'm in the minority, I'll revert the unroll, but I don't think a plan beyond "unroll them as I touch them" should be required given the existing inconsistency.

Reply:
Well, I've got quite the uphill battle ahead of me. I've reverted the unroll for now... soon.jpg

Reply:
We are inconsistent: some tests are in methods called from `run` and some are in `run` directly. I don't think we should settle on a single "hard" rule because such rules force us to do the wrong thing.

Consider that we have several tests that are several hundred lines long. Put all those in a single function and you'd end up with a function that's several thousand lines and that's just a recipe for disaster. Grouping a single set of "tests" inside a function just makes for easier to read code.

But then again, we have tests that are less than 10 lines long. Do we want a separate function for those? Probably not because it seems like an unnecessary level of indirection.

I think we need to exercise our judgement, and decide on a case-by-case basis. Yes, that means some inconsistency between files; Things like `Manifest_test` should probably be cleaned up to be one way or the other.

One thought might be to make `testcase` take both a name and a lambda. That way, we could develop a common style:

```
void run()
{
    testcase("Try one thing", []()
    {
    });

    testcase("Try a second thing", []()
    {
    });

    for (int i = 0; i != 8; ++i)
    {
        testcase ("Try various things: " + std::to_string(i), []()
        {
        }(i));
    }
}
```

At least until  we migrate to a different testing framework and away from `test_suite`; then we can see if one particular style is better suited.